# Lab 4
Working With Missing data

### Learning objectives
Upon successful completion of this lab, you will have demonstrated the abilities to:
1. Create missing values.
2. Recode and make calculations with missing values



In [1]:
import pandas as pd
from numpy import nan
import numpy as np

In [2]:
np.__version__

'1.26.4'

In [3]:
ebola = pd.read_csv('country_timeseries.csv')
ebola.dtypes
ebola.shape

(122, 18)

In [4]:
#print(ebola.count())
num_rows = ebola.shape[0]
num_missing = num_rows - ebola.count()
print(num_missing)

Date                     0
Day                      0
Cases_Guinea            29
Cases_Liberia           39
Cases_SierraLeone       35
Cases_Nigeria           84
Cases_Senegal           97
Cases_UnitedStates     104
Cases_Spain            106
Cases_Mali             110
Deaths_Guinea           30
Deaths_Liberia          41
Deaths_SierraLeone      35
Deaths_Nigeria          84
Deaths_Senegal         100
Deaths_UnitedStates    104
Deaths_Spain           106
Deaths_Mali            110
dtype: int64


1. Below is another way to count missing data in each of the columns of ebola dataset. Modify the code to display the percentage of missing values for each of the columns in ebola dataset.

In [5]:
num_rows = ebola.shape[0]
num_missing = num_rows - ebola.count()

missing_val_count_by_column = ebola.isnull().sum()
missing_percentages = missing_val_count_by_column / num_rows * 100
missing_percentages

Date                    0.000000
Day                     0.000000
Cases_Guinea           23.770492
Cases_Liberia          31.967213
Cases_SierraLeone      28.688525
Cases_Nigeria          68.852459
Cases_Senegal          79.508197
Cases_UnitedStates     85.245902
Cases_Spain            86.885246
Cases_Mali             90.163934
Deaths_Guinea          24.590164
Deaths_Liberia         33.606557
Deaths_SierraLeone     28.688525
Deaths_Nigeria         68.852459
Deaths_Senegal         81.967213
Deaths_UnitedStates    85.245902
Deaths_Spain           86.885246
Deaths_Mali            90.163934
dtype: float64

2. What is the percentage of missing values for each of the  columns in ebola dataset. Use a for loop and np.count_nonzero() function. Your output should show the column name and the percentage of missing values.  

In [6]:
ebola.columns # is a list of column names

Index(['Date', 'Day', 'Cases_Guinea', 'Cases_Liberia', 'Cases_SierraLeone',
       'Cases_Nigeria', 'Cases_Senegal', 'Cases_UnitedStates', 'Cases_Spain',
       'Cases_Mali', 'Deaths_Guinea', 'Deaths_Liberia', 'Deaths_SierraLeone',
       'Deaths_Nigeria', 'Deaths_Senegal', 'Deaths_UnitedStates',
       'Deaths_Spain', 'Deaths_Mali'],
      dtype='object')

In [7]:
np.count_nonzero(ebola['Deaths_Mali'].isnull())

110

In [8]:
for col in ebola.columns:
    print("% of data missing in {} column = {}%".format(col, missing_percentages[col]))

% of data missing in Date column = 0.0%
% of data missing in Day column = 0.0%
% of data missing in Cases_Guinea column = 23.770491803278688%
% of data missing in Cases_Liberia column = 31.967213114754102%
% of data missing in Cases_SierraLeone column = 28.688524590163933%
% of data missing in Cases_Nigeria column = 68.85245901639344%
% of data missing in Cases_Senegal column = 79.50819672131148%
% of data missing in Cases_UnitedStates column = 85.24590163934425%
% of data missing in Cases_Spain column = 86.88524590163934%
% of data missing in Cases_Mali column = 90.1639344262295%
% of data missing in Deaths_Guinea column = 24.59016393442623%
% of data missing in Deaths_Liberia column = 33.60655737704918%
% of data missing in Deaths_SierraLeone column = 28.688524590163933%
% of data missing in Deaths_Nigeria column = 68.85245901639344%
% of data missing in Deaths_Senegal column = 81.9672131147541%
% of data missing in Deaths_UnitedStates column = 85.24590163934425%
% of data missing in

3. Show the columns that have more than 90% missing values. Write a for loop on missing_val_count_by_column pandas Series from above. missing_val_count_by_column.index will give the list of column names and missing_val_count_by_column[i] will give the count of missing values where i is a column name.

In [9]:
print("Rows missing more than 90% of Data:")
for i in range(len(missing_val_count_by_column)):
  if missing_val_count_by_column[i] > 0.9 * num_rows:
    print(missing_val_count_by_column.index[i], missing_val_count_by_column[i]/num_rows*100)

Rows missing more than 90% of Data:
Cases_Mali 90.1639344262295
Deaths_Mali 90.1639344262295


<ipython-input-9-e8a6d23d7817>:3: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  if missing_val_count_by_column[i] > 0.9 * num_rows:
<ipython-input-9-e8a6d23d7817>:4: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  print(missing_val_count_by_column.index[i], missing_val_count_by_column[i]/num_rows*100)


4. Drop the columns that have more than 90% missing values.
        - define a list columns_to_drop containing the column names to drop
        - Create a new dataframe ebola_droppedCol
        - display the shape of the new dataframe.
        - print the number of missing values in ebola_droppedCol
        

In [10]:
columns_to_drop = []
# add columns with more than 90% missing values to list
for i in range(len(missing_val_count_by_column)):
  if missing_val_count_by_column[i] > 0.9 * num_rows:
    columns_to_drop.append(missing_val_count_by_column.index[i])

print(columns_to_drop)

['Cases_Mali', 'Deaths_Mali']


<ipython-input-10-569a4f191a93>:4: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  if missing_val_count_by_column[i] > 0.9 * num_rows:


In [11]:
ebola_droppedCol = ebola.drop(columns=columns_to_drop)
ebola_droppedCol.shape

(122, 16)

In [12]:
ebola_droppedCol.isnull().sum()

Date                     0
Day                      0
Cases_Guinea            29
Cases_Liberia           39
Cases_SierraLeone       35
Cases_Nigeria           84
Cases_Senegal           97
Cases_UnitedStates     104
Cases_Spain            106
Deaths_Guinea           30
Deaths_Liberia          41
Deaths_SierraLeone      35
Deaths_Nigeria          84
Deaths_Senegal         100
Deaths_UnitedStates    104
Deaths_Spain           106
dtype: int64

#### Fill Backward

5. Answer each of the following questions:
    - Count the number of missing values in the dataframe after the fill backward imputation of missing values. show the first 10 rows and 5 columns after imputation.
    - Explain what changes have been made after fill backward imputation of the missing values and why.


In [18]:
ebola.bfill().isnull().sum().sum()

732

In [19]:
ebola.iloc[0:10, 0:5].bfill()

Date  Day  Cases_Guinea  Cases_Liberia  Cases_SierraLeone
0    1/5/2015  289        2776.0         8166.0            10030.0
1    1/4/2015  288        2775.0         8166.0             9780.0
2    1/3/2015  287        2769.0         8166.0             9722.0
3    1/2/2015  286        2730.0         8157.0             9633.0
4  12/31/2014  284        2730.0         8115.0             9633.0
5  12/28/2014  281        2706.0         8018.0             9446.0
6  12/27/2014  280        2695.0         7977.0             9409.0
7  12/24/2014  277        2630.0         7977.0             9203.0
8  12/21/2014  273        2597.0         7862.0             9004.0
9  12/20/2014  272        2571.0         7862.0             8939.0

backfill means you are filling based on the NEXT known value. So for Liberia, first few values were null, first known value was index 2 - null values are replaced with the null.

#### Fill forward

6. Answer each of the following questions:
    - Count the number of missing values in the dataframe after the fill forward imputation of missing values. Show the first 10 rows and 5 columns after imputation.
    - Explain what changes have been made after fill forward imputation of the missing values and why.

In [20]:
ebola.ffill().isnull().sum().sum()

134

In [21]:
ebola.iloc[0:10, 0:5].ffill()

Date  Day  Cases_Guinea  Cases_Liberia  Cases_SierraLeone
0    1/5/2015  289        2776.0            NaN            10030.0
1    1/4/2015  288        2775.0            NaN             9780.0
2    1/3/2015  287        2769.0         8166.0             9722.0
3    1/2/2015  286        2769.0         8157.0             9722.0
4  12/31/2014  284        2730.0         8115.0             9633.0
5  12/28/2014  281        2706.0         8018.0             9446.0
6  12/27/2014  280        2695.0         8018.0             9409.0
7  12/24/2014  277        2630.0         7977.0             9203.0
8  12/21/2014  273        2597.0         7977.0             9004.0
9  12/20/2014  272        2571.0         7862.0             8939.0

Forward fill means it will fill with the LAST known value. Index 6 column Liberia was null, and after doing ffill(), is now 8018, since the last known value was 8018 at index 5.

7. Answer each of the following questions:
    - Replace the missing values with backward fill method, next apply forward fill method.
    - print the number of missing values after imputation.

In [22]:
double_fill = ebola.bfill().ffill()
double_fill.isnull().sum().sum()

0

8. Replace the missing values using a linear interpolation and print the top rows of the dataset.

In [23]:
interpolated = ebola.interpolate(method='linear')
interpolated.head()

<ipython-input-23-d230f2b44641>:1: FutureWarning: DataFrame.interpolate with object dtype is deprecated and will raise in a future version. Call obj.infer_objects(copy=False) before interpolating instead.
  interpolated = ebola.interpolate(method='linear')


Date  Day  Cases_Guinea  Cases_Liberia  Cases_SierraLeone  \
0    1/5/2015  289        2776.0            NaN            10030.0   
1    1/4/2015  288        2775.0            NaN             9780.0   
2    1/3/2015  287        2769.0         8166.0             9722.0   
3    1/2/2015  286        2749.5         8157.0             9677.5   
4  12/31/2014  284        2730.0         8115.0             9633.0   

   Cases_Nigeria  Cases_Senegal  Cases_UnitedStates  Cases_Spain  Cases_Mali  \
0            NaN            NaN                 NaN          NaN         NaN   
1            NaN            NaN                 NaN          NaN         NaN   
2            NaN            NaN                 NaN          NaN         NaN   
3            NaN            NaN                 NaN          NaN         NaN   
4            NaN            NaN                 NaN          NaN         NaN   

   Deaths_Guinea  Deaths_Liberia  Deaths_SierraLeone  Deaths_Nigeria  \
0         1786.0             NaN              2977.0             NaN   
1         1781.0             NaN              2943.0             NaN   
2         1767.0          3496.0              2915.0             NaN   
3         1753.0          3496.0              2871.0             NaN   
4         1739.0          3471.0              2827.0             NaN   

   Deaths_Senegal  Deaths_UnitedStates  Deaths_Spain  Deaths_Mali  
0             NaN                  NaN           NaN          NaN  
1             NaN                  NaN           NaN          NaN  
2             NaN                  NaN           NaN          NaN  
3             NaN                  NaN           NaN          NaN  
4             NaN                  NaN           NaN          NaN

In [ ]:
interpolated.isnull().sum()

Date                    0
Day                     0
Cases_Guinea            0
Cases_Liberia           2
Cases_SierraLeone       0
Cases_Nigeria          13
Cases_Senegal          13
Cases_UnitedStates     13
Cases_Spain            13
Cases_Mali             13
Deaths_Guinea           0
Deaths_Liberia          2
Deaths_SierraLeone      0
Deaths_Nigeria         13
Deaths_Senegal         13
Deaths_UnitedStates    13
Deaths_Spain           13
Deaths_Mali            13
dtype: int64

#### Short answer questions

9. Write a paragraph to highlight the differences between the fill foward/backward versus linear interpolation of the missing values.



Fill forward/backward method will cover all values, and therefore remove all nulls. The edge cases (ie. Beginning rows if the first null value is index 0, and the end rows if the last null value is index -1) are handled through backfill (which will fill with the NEXT known value, removing any nulls from the beginning), and forwardfill (which will fill with the LAST known value, removing any nulls from the end.

Linear interpolation, by default, will repeat the last known value for the end fo the dataframe, and will return NaN for the first NaN values at the top.

Linear Interpolation simply means to estimate a missing value by connecting dots in a straight line in increasing order. In short, It estimates the unknown value in the same increasing order from previous values. The default method used by Interpolation is Linear. So while applying it, we need not specify it. If there are no values to begin with (ie. if the first value is NaN), then there is to starting point to connect our dots from, causing a NaN value.

https://www.analyticsvidhya.com/blog/2021/06/power-of-interpolation-in-python-to-fill-missing-values/#:~:text=Linear%20Interpolation%20simply%20means%20to,we%20need%20not%20specify%20it.

10. Time series interpolation works by calling interpolate with the option method='time'.
    - Explain how is it different than the linear interpolation?
    - when do linear and time interpolation give the same results?

Linear interpolation will ignore the index, and assume that every datapoint is evenly spaced out.

Time-Series interpolation will not ignore the index, and will use it to determine how far apart the points are (ie. time series data would have a date as the index - it would assume if the index is separated by days, that we need to account for number of days between two non-missing values.)

They will therefore, give the same results if we have a complete index (ie. our time series data, has rows for every single day in a time period). Linear interpolation can ignore the index since it assumes all datapoints are equally spread out, and timeseries will know that all the datapoints actually ARE evenly spread out.